# Imports

* 무거운 "train" 작업 전체의 input의 output만 보고하는 튜토리얼 작성하고
* 그러고 나면 다음 단계로, train과정에서 들어가는 함수들을 쪼개서 분석해보는 tutorial도 해보기
  * `train.py`에서 호출하는 세부 모듈을 그대로 가져다 쓰되, train.py의 전체 흐름을 조금 더 가볍게 뫄하는 별도 스크립트 작성이 필요할 것으로 보임.


  train.py의 input이 뭔지 파악이 되고 나면 전체 프로그램의 흐름(모듈 간 의존관계)를 다이어그램(ppt)으로 그려보기
  * 핵심은, "특정 기능만 사용/분석 하려고 하면 뭐는 몰라도 되느냐"

  생각날 때 틈틈이:
     * 3DGS 발전 방향 고민 및 최근 연구 탐색(검색)해보기
     * multi-view 이미지를 3DGS 형태로 복원 --> mesh 형태로 변환하는

In [ ]:
import os
import torch
import numpy as np
import open3d as o3d
from random import randint
from utils.loss_utils import l1_loss, ssim
from gaussian_renderer import render, network_gui
import sys
from scene import Scene, GaussianModel
from utils.general_utils import safe_state, get_expon_lr_func
import uuid
from tqdm import tqdm
from utils.image_utils import psnr
from argparse import ArgumentParser, Namespace
from arguments import ModelParams, PipelineParams, OptimizationParams
from scene.dataset_readers import sceneLoadTypeCallbacks
try:
    from torch.utils.tensorboard import SummaryWriter
    TENSORBOARD_FOUND = True
except ImportError:
    TENSORBOARD_FOUND = False

try:
    from fused_ssim import fused_ssim
    FUSED_SSIM_AVAILABLE = True
except:
    FUSED_SSIM_AVAILABLE = False

try:
    from diff_gaussian_rasterization import SparseGaussianAdam
    SPARSE_ADAM_AVAILABLE = True
except:
    SPARSE_ADAM_AVAILABLE = False

Jupyter environment detected. Enabling Open3D WebVisualizer.
[Open3D INFO] WebRTC GUI backend enabled.
[Open3D INFO] WebRTCWindowSystem: HTTP handshake server disabled.


# Running train.py

We will run the following terminal command:

```cmd
python train.py -s Gaussiatest/Test2
```

Here this command assumes there are already point3D.ply which is the output of convet.py in [GaussianTest/Test2/sparse/0](GaussianTest/Test2/sparse/0)

# Input

`train.py`에 정의되어 있는 `training`이라는 함수에  

```python
# `train.py`
def training(...):
    ...
    gaussians = GaussianModel(dataset.sh_degree, opt.optimizer_type)  
    scene = Scene(dataset, gaussians) # Here
```

```python
# `scene/__init__.py`
class Scene:
    ...
    def __init__(self, args : ModelParams, gaussians : GaussianModel, load_iteration=None, shuffle=True, resolution_scales=[1.0]):
        ...
        if os.path.exists(os.path.join(args.source_path, "sparse")):
            scene_info = sceneLoadTypeCallbacks["Colmap"](args.source_path, args.images, args.depths, args.eval, args.train_test_exp) # Here
```

```python
# `scene/dataset_readers`
sceneLoadTypeCallbacks = { # Line 312
    "Colmap": readColmapSceneInfo, # Here
    "Blender" : readNerfSyntheticInfo
}

def readColmapSceneInfo(path, images, depths, eval, train_test_exp, llffhold=8): # Line 145
    try:
        cameras_extrinsic_file = os.path.join(path, "sparse/0", "images.bin")
        cameras_intrinsic_file = os.path.join(path, "sparse/0", "cameras.bin")
        cam_extrinsics = read_extrinsics_binary(cameras_extrinsic_file)
        cam_intrinsics = read_intrinsics_binary(cameras_intrinsic_file)
    except:
        cameras_extrinsic_file = os.path.join(path, "sparse/0", "images.txt")
        cameras_intrinsic_file = os.path.join(path, "sparse/0", "cameras.txt")
        cam_extrinsics = read_extrinsics_text(cameras_extrinsic_file)
        cam_intrinsics = read_intrinsics_text(cameras_intrinsic_file)

    depth_params_file = os.path.join(path, "sparse/0", "depth_params.json")
    ...

    cam_infos_unsorted = readColmapCameras(...) # returns CameraInfo objects
    cam_infos = sorted(cam_infos_unsorted.copy(), key = lambda x : x.image_name)

    train_cam_infos = [c for c in cam_infos if train_test_exp or not c.is_test] # scene_info's train_cameras
    test_cam_infos = [c for c in cam_infos if c.is_test] # scene_info's test_cameras
    ...
    ply_path = os.path.join(path, "sparse/0/points3D.ply") # scene_info's ply_path
    ...
    pcd = fetchPly(ply_path) # scene_info's point_cloud
    ...

    scene_info = SceneInfo(point_cloud=pcd,
                           train_cameras=train_cam_infos,
                           test_cameras=test_cam_infos,
                           nerf_normalization=nerf_normalization,
                           ply_path=ply_path,
                           is_nerf_synthetic=False)
    return scene_info
```



pcd를 불러오기 위한 `fetchply` 함수
```python
    def fetchPly(path):
        plydata = PlyData.read(path)
        vertices = plydata['vertex']
        positions = np.vstack([vertices['x'], vertices['y'], vertices['z']]).T
        colors = np.vstack([vertices['red'], vertices['green'], vertices['blue']]).T / 255.0
        normals = np.vstack([vertices['nx'], vertices['ny'], vertices['nz']]).T
        return BasicPointCloud(points=positions, colors=colors, normals=normals)
```



input.ply가 생성되는 부분
```python
# scene.__init__.py
# Line 51
if not self.loaded_iter:
    with open(scene_info.ply_path, 'rb') as src_file, open(os.path.join(self.model_path, "input.ply") , 'wb') as dest_file:
        dest_file.write(src_file.read())

```

위와 같은 코드가 있는데 `Scene` 클래스에서 `scene.dataset_readers`의 `sceneLoadTypeCallbacks["Colmap"]`라는 메소드로 [tutorial_convert_py.ipynb](tutorial_convert_py.ipynb)에서 했던 것과 같이 `Image Undistorter`가 진행된 후의 images.bin 파일과 cameras.bin 파일을 읽어온다. 이 값들을 `scene_info` 변수가 가져오고 `prepare_output_and_logger` 메소드가 설정한 [args.model_path](output/Test2_Result)에 `input.ply`(새로 만들어짐) 파일로 [GaussianTest/Test2/sparse/0](GaussianTest/Test2/sparse/0)/point3D.ply의 값을 복사하여 가져온다. 이를 GaussianModel 클래스의 `create_from_pcd` 메소드를 이용해 3D Gaussian을 만든다.

## Input.ply

In [2]:
ply_path = 'output/Test2_Result/input.ply'
pcd = o3d.io.read_point_cloud(ply_path)

print(pcd)
points = np.asarray(pcd.points)
print(points)

o3d.visualization.draw_geometries([pcd])

PointCloud with 1768 points.
[[  4.34101534  -1.97590983   8.94548512]
 [  6.54583645  10.24704552   2.88442421]
 [  4.26116467  -0.91624063   9.97420979]
 ...
 [-14.37703228  10.57718468   7.03714371]
 [ -3.41339111   4.91355276  12.40224266]
 [-14.48631001  10.56289577   6.95793295]]


# create_from_pcd

`Scene` 클래스 객체를 생성하는데 있어 마지막으로 실행하는 함수이다. 이 함수가 무엇을 하는지 알아보고자 한다. 

```python
# scene.__init__.py
self.gaussians.create_from_pcd(scene_info.point_cloud, scene_info.train_cameras, self.cameras_extent) # Line 83


# scene.gaussian_model.py
def create_from_pcd(self, pcd : BasicPointCloud, cam_infos : int, spatial_lr_scale : float):
```

## Parameters

`pcd:` input에서 봤던 SceneInfo 객체의 point_cloud 값을 가져온다. 이는 BasicPointCloud 클래스의 객체이고 input.ply의 내용이 담겨져 있다.  
```python
BasicPointCloud(points=positions, colors=colors, normals=normals)

```

`cam_infos:` SceneInfo 객체의 train_cameras의 값을 가져온다. train_cameras는 `readColmapCameras`의 return 값인 `CameraInfo` 객체를 가지고 있다. 이 파라미터는 코드 내에서 실제로는 len(cam_infos)로만 사용되기에 `int`로 처리되었다.
```python
# scene.dataset_readers.py Line 194
cam_infos_unsorted = readColmapCameras(...) # returns CameraInfo objects
cam_infos = sorted(cam_infos_unsorted.copy(), key = lambda x : x.image_name) 

train_cam_infos = [c for c in cam_infos if train_test_exp or not c.is_test] # scene_info's train_cameras

```
`spatial_lr_scale:` SceneInfo 객체의 `nerf_normalization[radius]` 값을 가져오는데 이는 `getNerfppNorm`이라는 함수가 SceneInfo 객체의 `train_cam_infos`에 속한 다수의 카메라 좌표들을 구해 중심을 구하고 그 중심에서 가장 먼 카메라와의 거리를 이용해 정규화를 시켜 씬의 전체 크기를 가늠하는 기준 scale 값으로 사용한다.
```python
# scene.dataset_readers.py Line 203
nerf_normalization = getNerfppNorm(train_cam_infos)

def getNerfppNorm(cam_info):
    ...
    return {"translate": translate, "radius": radius}
```

## How it works

```python
self.spatial_lr_scale = spatial_lr_scale

```
`GaussianModel` 클래스의 멤버 변수인 `spatial_lr_scale`에 파라미터로 받아온 `self.camera_extent` 값을 할당한다.
```python


"""Recall
class BasicPointCloud(NamedTuple):
    points : np.array
    colors : np.array
    normals : np.array
"""

fused_point_cloud = torch.tensor(np.asarray(pcd.points)).float().cuda()
fused_color = RGB2SH(torch.tensor(np.asarray(pcd.colors)).float().cuda())
```
`fused_point_cloud` 변수에 `pcd`의 `points`에 해당하는 값들을 float형 array로 받아오고 이를 GPU에서 계산할 수 있도록 한다.  
`fused_color` 변수에 `pcd`의 `corlors`에 해당하는 값들을 SH의 초기 계수로 변환하여 float형 array로 받아오고 이를 GPU에서 계산할 수 있도록 한다.

**학습을 더 용이하게 하기 위하여 RGB 값을 SH의 초기 계수로 변환하는 과정을 거친다.**
```python

features = torch.zeros((fused_color.shape[0], 3, (self.max_sh_degree + 1) ** 2)).float().cuda()
features[:, :3, 0 ] = fused_color
features[:, 3:, 1:] = 0.0
```
`features` 변수에 0으로 차 있는 `fused_color.shape[0]`개의 `3x16 행렬` ($\because$ max_sh_degree $== 3$) 을 만든다.  
각 `3x16 행렬`의 처음 3행의 첫 번째 열의 값을 RGB 값으로 채우고 나머지는 전부 0으로 한다. 
```python

dist2 = torch.clamp_min(distCUDA2(torch.from_numpy(np.asarray(pcd.points)).float().cuda()), 0.0000001)
scales = torch.log(torch.sqrt(dist2))[...,None].repeat(1, 3)
rots = torch.zeros((fused_point_cloud.shape[0], 4), device="cuda")
rots[:, 0] = 1

opacities = self.inverse_opacity_activation(0.1 * torch.ones((fused_point_cloud.shape[0], 1), dtype=torch.float, device="cuda")) # Here

# scene.gaussian_model.py Line 45
self.inverse_opacity_activation = inverse_sigmoid

# utils.general_utils.py Line 18
def inverse_sigmoid(x):
    return torch.log(x/(1-x))
```
`dist2:` 가장 가까운 점과의 거리의 제곱을 0.0000001보다 크게 하여 dist2에 tensor로 저장.  
`scales:` 루트를 씌워 거리를 구한다.  
`rots:` point cloud의 개수에 대해 `quaternion`의 `w` 값을 1로 함으로써 각 점에 대해 아무런 회전도 하지 않은 상태로 만든다.  
`opacities:` 1로 채워진 nx1 행렬을 만들어 `inverse_sigmoid`에 넣는다.

# Outputs

By running train.py its ouputs are stored in directory [output/Test2_Result](output/Test2_Result)

### Details

Information [input.ply](output/Test2_Result/input.ply) have is the result of running convert.py. Which is 1768 point clouds.

## Results for 7000 iterations

In [4]:
ply_path = 'output/Test2_Result/point_cloud/iteration_7000/point_cloud.ply'
pcd = o3d.io.read_point_cloud(ply_path)

print(pcd)
points = np.asarray(pcd.points)
print(points)

o3d.visualization.draw_geometries([pcd])

PointCloud with 347527 points.
[[ 6.59872437 10.12901211  2.86574125]
 [ 0.22425154  1.65513897 12.2174387 ]
 [ 4.22987652 -1.58260691  9.34566116]
 ...
 [ 0.07556491  6.75840139  6.57672215]
 [ 4.2516675   7.14003801 10.84717178]
 [-0.48998436  5.14745998  7.08731127]]


## Results for 30000 iterations

In [3]:
ply_path = 'output/Test2_Result/point_cloud/iteration_30000/point_cloud.ply'
pcd = o3d.io.read_point_cloud(ply_path)

print(pcd)
points = np.asarray(pcd.points)
print(points)

o3d.visualization.draw_geometries([pcd])

PointCloud with 375184 points.
[[ 6.60873985 10.11532307  2.88246822]
 [ 0.17242943  1.56833577 12.10415363]
 [ 4.23568726 -1.58306515  9.3474226 ]
 ...
 [-1.06505299  5.10284138 12.32131481]
 [ 1.89028597  3.38411283 10.96808529]
 [-0.26036382  4.30767822 11.26516533]]
